cli/openApiSpec.ts

```diff
@@ -0,0 +1,276 @@
+    const [modulePath, functionName] = name.split(':')
+    const withoutExtension = modulePath.slice(0, modulePath.length - 3)
+    const pathParts = withoutExtension.split('/')
+    const shortName = `${pathParts.join('.')}.${functionName}`
+    const urlPathName = `${pathParts.join('/')}/${functionName}`
+    return {
+        original: name,
+        shortName,
+        urlPathName,
+    }
+}
+
+function generateEndpointDef(func: AnalyzedFunction) {
+    const { urlPathName, shortName } = formatName(func.identifier)
+    return `
+    /api/run/${urlPathName}:
+      post:
+        summary: Calls a ${func.functionType.toLowerCase()} at the path ${func.identifier}
+        tags: 
+          - ${func.functionType.toLowerCase()}
+        requestBody:
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Request_${shortName}'
+          required: true
+        responses:
+          '200':
+            description: Convex executed your request and returned a result
+            content:
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/Response_${shortName}'    
+          '400':
+            description: Failed operation
+            content:
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/FailedResponse'
+          '500':
+            description: Convex Internal Error
+            content:
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/FailedResponse'\n`
+}
+
+function generateEndpointSchemas(func: AnalyzedFunction) {
+    const { shortName } = formatName(func.identifier)
+    return `
+    Request_${shortName}:
+      type: object
+      required:
+        - args
+      properties:
+        args:\n${reindent(
+        generateSchemaFromValidator(func.args ?? { type: 'any' }),
+        5
+    )}\n
+    Response_${shortName}:
+      type: object
+      required:
+        - status
+      properties:
+        status: 
+          type: string
+          enum:
+            - "success"
+            - "error"
+        errorMessage:
+          type: string
+        errorData:
+          type: object
+        value:\n${reindent(
+        generateSchemaFromValidator(func.returns ?? { type: 'any' }),
+        5
+    )}\n`
+}
+
+export function generateOpenApiSpec(functionSpec: FunctionSpec) {
+    return `
+openapi: 3.0.3
+info:
+    title: Convex App - OpenAPI 3.0
+    version: 0.0.0
+servers:
+    - url: ${functionSpec.url}
+security:
+  - bearerAuth: []
+tags:
+    - name: query
+      description: Functions that read data
+    - name: mutation
+      description: Functions that write/update/delete data
+    - name: action
+      description: Functions that can make calls to external APIs
+paths:
+${reindent(
+        // Skip http actions because they go to a different url and we don't have argument/return types
```

Comment:
Should we also filter out internal functions? Or potentially document somewhere that the internal ones need an admin key in the auth header?

Reply:
i'll add this as a flag. we should definitely document this tho